### ISSUE TITLE
AArch64 blake3 assembly trashes x18

### ISSUE THREAD

**Issue by zxombie** [openzfs/zfs#14714]
### System information

Type | Version/Name
 --- | ---
Distribution Name	| FreeBSD
Distribution Version	| 14.0-CURRENT
Kernel Version	| 14.0-CURRENT
Architecture	| arm64
OpenZFS Version	| zfs-2.1.99-FreeBSD_g57cfae4a2

### Describe the problem you're observing

Kernel panics because the accelerated blake3 implementation trashes the x18 register. In `b3_aarch64_sse2.S` and `b3_aarch64_sse41.S` the x18 register is used. This is reserved on FreeBSD/arm64 for a per-CPU pointer. The kernel is built with the `-ffixed-x18` compiler flag to ensure C code doesn't write to x18, however the above two files modify either it or the 32 bit alias w18.

### Describe how to reproduce the problem

Boot on FreeBSD/arm64 with the zfs module

**Comment by kevans91** [comment]
> I will fix this and provide a pull request soon.

awesome, thanks! I have another patch for more freebsd/aarch64 fallout that I'll push through in a PR momentarily as well to complement yours. freebsd/arm is also going to be broken for similar reasons (needed kfpu_ handling), but I don't really have a good setup to test that with and I'm not sure how many people use it on armv7 anyways...

**Comment by zxombie** [comment]
Is this generated from the blake3 C code? If so would it be a problem to also build with `-mbranch-protection=standard`? It adds support for the pointer authentication and branch target identification extensions. The new instructions it adds are nops in CPUs that don't implement these extensions.

**Comment by rincebrain** [comment]
I have a working FreeBSD/AArch64 with the various SIMD acceleration modes implemented that passed everything last I tried it, as I recall, I should probably push that...

**Comment by kevans91** [comment]
I pushed part of my patch in PR #14715, just to implement kfpu_begin/kfpu_end. Combined with a hackaround for blake2 to avoid x18, that was sufficient to boot again, but this system just uses `checksum=on` everywhere and thus isn't fully exercising much beyond the initial benchmarking that was breaking the system.

**Comment by lundman** [comment]
I'd be curious to test and see if this is related to the aarch64 on macOS issues.

**Comment by kevans91** [comment]
> I'd be curious to test and see if this is related to the aarch64 on macOS issues.

fwiw, I noted in one of my emails that this probably is broken for the same reason on macOS. See: https://developer.apple.com/documentation/xcode/writing-arm64-code-for-apple-platforms

> The platforms reserve register x18. Don’t use this register.

**Comment by lundman** [comment]
I wonder if `CC_AArch64Apple` ever landed in llvm and I can use it for this as well.

**Comment by oconnor663** [comment]
> Is this generated from the blake3 C code?

Reading the comments at the top of those assembly files, it looks like they're converted from upstream x86 assembly, using a tool called SIMD Everywhere.

**Comment by kevans91** [comment]
> > Is this generated from the blake3 C code?
> 
> Reading the comments at the top of those assembly files, it looks like they're converted from upstream x86 assembly, using a tool called SIMD Everywhere.

Yeah, we've since worked out how to generate them and from where, but unfortunately the process is incredibly undocumented and makes this a very unreproducible process. At the very least:

1. What compiler flags should be used? I guess `-O2` at a minimum + the ones we're asking for, but it's undocumented
2. To what extent is the output massaged? I see at least symbols have been renamed, what else has happened?

Knowing what compiler version was used would also be helpful to try and avoid major changes in codegen. These are the kinds of things that would've been nice to see documented in the output files or the commit message so that others could reproduce the massive assembly blobs.

**Comment by oconnor663** [comment]
If I understand the question correctly, the answer is that @sneves wrote upstream assembly files by hand, so they aren't reproducible from the output of any compiler.

**Comment by kevans91** [comment]
> If I understand the question correctly, the answer is that @sneves wrote upstream assembly files by hand, so they aren't reproducible from the output of any compiler.

These specific files are versions of `blake3_sse*.c` compiled into assembly and pushed into the openzfs tree; upstream only has x86 versions of these in assembly.

**Comment by rincebrain** [comment]
It's not called _neon_impl because there is a _neon_impl in the BLAKE3 tree and it's slow, it's called the same thing because it's compiled from the SSE intrinsics implementations with SIMDe, which implements compiler intrinsics for platform A in terms of platform B's.

**Comment by bsdimp** [comment]
> It's not called _neon_impl because there is a _neon_impl in the BLAKE3 tree and it's slow, it's called the same thing because it's compiled from the SSE intrinsics implementations with SIMDe, which implements compiler intrinsics for platform A in terms of platform B's.

It should have comments about why this is done then. Currently it's all a mass of undocumented tribal knowledge that's confusing for people that aren't well versed in the history of why this came to be.

**Comment by rincebrain** [comment]
Sure, I'm almost always in favor of more documentation.

**Comment by mcmilk** [comment]
I think this issue is fixed and can be closed - is this correct?

**Comment by mcmilk** [comment]
@tonyhutter - can you close this issue? It's fixed.
